### PR TITLE
2.0.14: handle pinned publishers before statistical voting

### PR DESCRIPTION
**Pull request by mrose17**
#given 

    const weights = [
      { publisher: '10%', weight: 0.10, pinPercentage: 10 },
      { publisher: '20%', weight: 0.20, pinPercentage: 20 },
      { publisher: '30%', weight: 0.30, pinPercentage: 30 },
      { publisher: '25%', weight: 0.25, pinPercentage: 25 },
      { publisher: '15%', weight: 0.15, pinPercentage: 15 }
    ]

i (now) get

[  "10%",  "20%",  "20%",  "30%",  "30%",  "30%",  "25%",  "25%",  "15%",  "15%" ]

### CODE REVIEW

File: index.js
Comment (NejcZdovc) [inline index.js]: can we do `if ((result.pinPercentage == null) || (result.pinPercentage <= 0)) return` so that we check null as well. Just in case

File: index.js
Comment (NejcZdovc) [inline index.js]: this can be `const`

File: index.js
Comment (mrose17) [reply index.js]: +1

File: index.js
Comment (mrose17) [reply index.js]: +1

**Comment by NejcZdovc** [comment]
looks good to me. Let me try it in the browser as well before we merge it

File: index.js
Comment (diracdeltas) [inline index.js]: can just do `const winners = []` and delete https://github.com/brave-intl/bat-publisher/pull/24/files#diff-168726dbe96b3ce427e7fedce31bb0bcR370

**Comment by diracdeltas** [comment]
this should have a unit test for the case where all publishers are pinned, which IIUC should be easy to do since the output is deterministic (non-randomized).

File: index.js
Comment (mrose17) [reply index.js]: good catch!

**Comment by mrose17** [comment]
i agree that it **should** have a unit test, but as with https://github.com/brave-intl/bat-ledger/pull/158 i must plead a lack of resources. however, the description above shows what my informal testing yields...

**Comment by riastradh-brave** [comment]
@mrose17 I don't understand what the semantics of the new `winners` function is supposed to be.  Can you summarize in English?  Why does your test case show two entries with `30%`, but three with `25%`?

The `winner` function and the `README.md` file suggest that `winners` is supposed to return an independent random sample from some distribution on votes, but your new `winners` function doesn't do that.  If someone called `winner` over and over again, they'd get the same result each time under certain choices of pinned weights, with your new code.

**Comment by mrose17** [comment]
in plain english: the `winner` function takes the weighted list of publishers and generates an array of votes that should be generated given that weight.

the current version of the function didn't treat pinned publishers any differently than un-pinned publishers, e.g., if `wsj.com` was pinned at `20%`, then it had a one-in-five chance for each vote of being selected. the people have spoken and they do not like being surprised with that. so, the new code gives pinned publishers the closest number of integer votes based on the percentage. any votes that remain then go through statistical voting.

the above example is deterministic, which is what @diracdeltas asked for an example of.

**Comment by mrose17** [comment]
@riastradh-brave - note that until this is integrated with the browser code (to use the new published version) and that the `ledger.js` middleware in the browser includes the pinned percentage in the `weights` array, the behavior of the existing code and new code will be same (statistical voting) only.

**Comment by NejcZdovc** [comment]
>and that the ledger.js middleware in the browser includes the pinned percentage in the weights array

we already set weight based on the pinned value @mrose17

**Comment by mrose17** [comment]
@NejcZdovc - but the problem is knowing whether the weight is due to pinning or not, hence the need to include the `pinPercentage` property, if set.

**Comment by NejcZdovc** [comment]
@mrose17 which we already do
https://github.com/brave/browser-laptop/blob/9d5977ae9c9965dc9a9060426c60f15fbee3d73f/app/browser/api/ledger.js#L530

Do we need to do anything else?

**Comment by mrose17** [comment]
@riastradh-brave - you asked

         Can you address why there are two votes for the 30% publisher but three for the 25% one?

the answer is: both 25% and 30% result in 3 votes (`Math.round`), but the call to `underscore.max` doesn't find the correct entry to trim (11 votes are generated instead of 10 because of the rounding).  it does now.

good catch!

**Review by riastradh-brave — DISMISSED** [review]
I am pretty sure I understand what this is doing now and it seems to do what is intended, modulo a minor rounding issue which was already there (https://github.com/brave-intl/bat-publisher/issues/25).  I created a separate issue to document the intent of the procedure: https://github.com/brave-intl/bat-publisher/issues/26

**Comment by mrose17** [comment]
@riastradh-brave - thanks!

**Comment by diracdeltas** [comment]
@riastradh-brave or @mrose17, could you please add @riastradh-brave's simple test from https://github.com/brave-intl/bat-publisher/pull/27 to this PR? thanks

also in case it wasn't obvious, Taylor has to be assigned as a reviewer in order for his approval to work.

**Comment by mrose17** [comment]
@diracdeltas - i would prefer that @riastradh-brave add it to this PR.

is @riastradh-brave taylor? if so, i see him above.

**Comment by mrose17** [comment]
@riastradh-brave - cf., https://github.com/brave-intl/bat-publisher/pull/24/files?diff=split#diff-168726dbe96b3ce427e7fedce31bb0bcR370 regarding why not passing `n` results in 1 result being returned... `.winners()` gives you exactly 1 winner.

**Comment by mrose17** [comment]
@riastradh-brave - are you blocked on me for this? i asked earlier for you to merge the test into this branch. any progress on that?

**Review by riastradh-brave — APPROVED** [review]
Approve for now, pending a proper reconsideration of the algorithm as several of us discussed privately.

**Pull request merged**
